ipykernel/debugger.py

```diff
@@ -562,6 +564,24 @@ async def richInspectVariables(self, message):
         reply["success"] = True
         return reply
 
+    async def modules(self, message):        
+        modules = [sys.modules[name] for name in sys.modules]
+        mods = []
+        for module in modules:
+            m = str(module)
```

Comment:
Are you looking for `mod.__spec__.origin`  ? (`__spec__` may be non so you would need `getattr(..., '__spec__', None)`, so not sure it is easier.

Reply:
I came  up with the following which is a bit less regexp-y:

```python
    async def modules(self, message):        
        modules = sys.modules.values()
        mods = []
        for module in modules:
            filename = getattr(getattr(module, "__spec__", None), "origin", None)
            if filename and filename.endswith(".py"):
                mods.append({mod.__name__: filename})

        reply = {"body": {"modules": mods, "totalModules": len(modules)}}
        return reply
```

One thing which is not clear to me is why would `mod.__name__` be different than the key in `sys.modules`, and is that roughly trying to do:

```
mods = [{k:m.__spec__.origin} for k,m in sys.modules.items()]
```

Reply:
I have implemented your suggestion in 8777e68e56712fef254ea72c70dca147b6e56a6d

Not sure how `sys.modules` should behave. I came to that code block in an empirical way.